File: python/rafter_cli/commands/upgrade.py
"""Upgrade command — upgrade rafter to the latest version."""
from __future__ import annotations

import os
import subprocess
import sys
from typing import Optional

import typer
from rich import print as rprint

from .. import __version__
from ..utils.formatter import fmt

upgrade_app = typer.Typer(name="upgrade", invoke_without_command=True, no_args_is_help=False)
update_app = typer.Typer(name="update", invoke_without_command=True, no_args_is_help=False)

PYPI_URL = "https://pypi.org/pypi/rafter-cli/json"
PYPI_PACKAGE = "rafter-cli"


def _is_ci() -> bool:
    return any(
        os.environ.get(v)
        for v in (
            "CI",
            "CONTINUOUS_INTEGRATION",
            "GITHUB_ACTIONS",
            "GITLAB_CI",
            "CIRCLECI",
            "TRAVIS",
            "JENKINS_URL",
        )
    )


def _is_newer(current: str, latest: str) -> bool:
    def parts(v: str):
        return [int(x) for x in v.split(".")]
    try:
        return parts(latest) > parts(current)
    except ValueError:
        return False


def _fetch_latest_version() -> str:
    import urllib.request
    import json

    with urllib.request.urlopen(PYPI_URL, timeout=5) as resp:
        data = json.loads(resp.read())
    return data["info"]["version"]


def _detect_installer() -> Optional[str]:
    """Return the install command prefix if detectable, else None."""
    executable = sys.executable

    # pipx: executable lives under PIPX_HOME or ~/.local/pipx
    pipx_home = os.environ.get("PIPX_HOME") or os.path.expanduser("~/.local/pipx")
    if executable.startswith(pipx_home) or "pipx" in executable:
        return "pipx"

    # uv virtual env: check for uv.lock in cwd or VIRTUAL_ENV parent
    virtual_env = os.environ.get("VIRTUAL_ENV", "")
    if virtual_env:
        parent = os.path.dirname(virtual_env)
        if os.path.exists(os.path.join(parent, "uv.lock")):
            return "uv"
        if os.path.exists(os.path.join(parent, "poetry.lock")):
            return "poetry"

    # uv tool installs: executable contains "uv" or ".uv"
    if "uv/tools" in executable or os.path.join(".uv", "tools") in executable:
        return "uv"

    return None


def _build_upgrade_command(installer: Optional[str]) -> list[str]:
    if installer == "pipx":
        return [["pipx", "upgrade", PYPI_PACKAGE]]
    if installer == "uv":
        return [["uv", "tool", "upgrade", PYPI_PACKAGE]]
    if installer == "poetry":
        return [["poetry", "add", f"{PYPI_PACKAGE}@latest"]]
    # pip fallback (may need pip3 on some systems)
    return [
        [sys.executable, "-m", "pip", "install", "--upgrade", PYPI_PACKAGE],
    ]


def _upgrade_main(
    check: bool = typer.Option(False, "--check", help="Check for updates without installing"),
    yes: bool = typer.Option(False, "--yes", "-y", help="Run the upgrade command automatically"),
) -> None:
    """Upgrade rafter to the latest version."""
    if _is_ci():
        rprint(fmt.info("CI environment detected — skipping upgrade check."))
        raise typer.Exit(0)

    try:
        latest_version = _fetch_latest_version()
    except Exception:
        rprint(fmt.error("Could not reach PyPI. Check your network connection."), file=sys.stderr)
        raise typer.Exit(1)

    if check:
        typer.echo(latest_version)
        return

    rprint(fmt.info(f"Current version: {__version__}"))
    rprint(fmt.info(f"Latest version:  {latest_version}"))
    rprint()

    if not _is_newer(__version__, latest_version):
        rprint(fmt.success("Already up to date."))
        return

    installer = _detect_installer()
    commands = _build_upgrade_command(installer)

    if yes:
        cmd = commands[0]
        rprint(fmt.info(f"Running: {' '.join(cmd)}"))
        rprint()
        result = subprocess.run(cmd)
        if result.returncode != 0:
            rprint(fmt.error("Upgrade failed. Try running the command manually:"))
            rprint(f"  {' '.join(cmd)}")
            raise typer.Exit(1)
        rprint()
        rprint(fmt.success(f"Upgraded to {latest_version}"))
        return

    # Print the command(s) to run
    if len(commands) == 1 and installer is not None:
        rprint(fmt.info("Run to upgrade:"))
        rprint(f"  {' '.join(commands[0])}")
    else:
        rprint(fmt.info("Run one of the following to upgrade:"))
        rprint(f"  pip install --upgrade {PYPI_PACKAGE}")
        rprint(f"  pipx upgrade {PYPI_PACKAGE}")
        rprint(f"  uv tool upgrade {PYPI_PACKAGE}")
        rprint()
        rprint(fmt.info("Or for a npm/pnpm installation:"))
        rprint("  npm install -g @rafter-security/cli@latest")

    rprint()


upgrade_app.callback()(_upgrade_main)
update_app.callback()(_upgrade_main)

File: node/src/commands/upgrade.ts
import { Command } from "commander";
import { execSync } from "child_process";
import axios from "axios";
import { fmt } from "../utils/formatter.js";
import { createRequire } from "module";

const require = createRequire(import.meta.url);
const { version: CURRENT_VERSION } = require("../../package.json");

const NPM_PACKAGE = "@rafter-security/cli";
const NPM_REGISTRY_URL = `https://registry.npmjs.org/${NPM_PACKAGE}/latest`;

function isCI(): boolean {
  return !!(
    process.env.CI ||
    process.env.CONTINUOUS_INTEGRATION ||
    process.env.GITHUB_ACTIONS ||
    process.env.GITLAB_CI ||
    process.env.CIRCLECI ||
    process.env.TRAVIS ||
    process.env.JENKINS_URL
  );
}

function isNewer(current: string, latest: string): boolean {
  const c = current.split(".").map(Number);
  const l = latest.split(".").map(Number);
  for (let i = 0; i < 3; i++) {
    if ((l[i] || 0) > (c[i] || 0)) return true;
    if ((l[i] || 0) < (c[i] || 0)) return false;
  }
  return false;
}

async function fetchLatestVersion(): Promise<string> {
  const res = await axios.get(NPM_REGISTRY_URL, { timeout: 5000 });
  return res.data.version as string;
}

type PackageManager = "npm" | "pnpm" | "yarn" | "unknown";
type InstallScope = "global" | "local" | "unknown";

interface InstallContext {
  manager: PackageManager;
  scope: InstallScope;
}

function detectInstallContext(): InstallContext {
  // npm_config_user_agent is set by npm, pnpm, and yarn when running scripts
  const userAgent = process.env.npm_config_user_agent || "";
  let manager: PackageManager = "unknown";
  if (userAgent.startsWith("pnpm/")) {
    manager = "pnpm";
  } else if (userAgent.startsWith("yarn/")) {
    manager = "yarn";
  } else if (userAgent.startsWith("npm/")) {
    manager = "npm";
  }

  // Fallback: check PNPM_HOME (set by pnpm global installs)
  if (manager === "unknown" && process.env.PNPM_HOME) {
    manager = "pnpm";
  }

  // Check executable path for clues
  const binPath = process.argv[1] ?? "";
  if (manager === "unknown") {
    if (binPath.includes("pnpm") || binPath.includes(".pnpm")) {
      manager = "pnpm";
    } else if (binPath.includes("yarn")) {
      manager = "yarn";
    }
  }

  // Scope detection
  let scope: InstallScope = "unknown";
  if (process.env.npm_config_global === "true") {
    scope = "global";
  } else if (binPath.includes("node_modules")) {
    scope = "local";
  } else {
    // Binary outside node_modules → likely global
    scope = "global";
  }

  return { manager, scope };
}

function buildUpgradeCommand(ctx: InstallContext): string | null {
  const pkg = `${NPM_PACKAGE}@latest`;
  if (ctx.manager === "unknown") return null;

  switch (ctx.manager) {
    case "npm":
      return ctx.scope === "local"
        ? `npm install ${pkg}`
        : `npm install -g ${pkg}`;
    case "pnpm":
      return ctx.scope === "local"
        ? `pnpm add ${pkg}`
        : `pnpm add -g ${pkg}`;
    case "yarn":
      return ctx.scope === "local"
        ? `yarn add ${pkg}`
        : `yarn global add ${pkg}`;
  }
}

export function createUpgradeCommand(): Command {
  const cmd = new Command("upgrade")
    .alias("update")
    .description("Upgrade rafter to the latest version")
    .option("--check", "Check for updates without installing")
    .option("-y, --yes", "Run the upgrade command automatically without prompting")
    .action(async (opts: { check?: boolean; yes?: boolean }) => {
      // No-op in CI
      if (isCI()) {
        console.log(fmt.info("CI environment detected — skipping upgrade check."));
        process.exit(0);
      }

      let latestVersion: string;
      try {
        latestVersion = await fetchLatestVersion();
      } catch {
        console.error(fmt.error("Could not reach npm registry. Check your network connection."));
        process.exit(1);
      }

      if (opts.check) {
        console.log(latestVersion);
        return;
      }

      console.log(fmt.info(`Current version: ${CURRENT_VERSION}`));
      console.log(fmt.info(`Latest version:  ${latestVersion}`));
      console.log();

      if (!isNewer(CURRENT_VERSION, latestVersion)) {
        console.log(fmt.success("Already up to date."));
        return;
      }

      const ctx = detectInstallContext();
      const upgradeCmd = buildUpgradeCommand(ctx);

      if (upgradeCmd && opts.yes) {
        console.log(fmt.info(`Running: ${upgradeCmd}`));
        console.log();
        try {
          execSync(upgradeCmd, { stdio: "inherit" });
          console.log();
          console.log(fmt.success(`Upgraded to ${latestVersion}`));
        } catch {
          console.error(fmt.error("Upgrade failed. Try running the command manually:"));
          console.log(`  ${upgradeCmd}`);
          process.exit(1);
        }
        return;
      }

      // Print the command(s) to run
      if (upgradeCmd) {
        console.log(fmt.info("Run to upgrade:"));
        console.log(`  ${upgradeCmd}`);
      } else {
        console.log(fmt.info("Run one of the following to upgrade:"));
        console.log(`  npm install -g ${NPM_PACKAGE}@latest`);
        console.log(`  pnpm add -g ${NPM_PACKAGE}@latest`);
        console.log(`  yarn global add ${NPM_PACKAGE}@latest`);
        console.log();
        console.log(fmt.info("Or for a local install:"));
        console.log(`  npm install ${NPM_PACKAGE}@latest`);
      }

      console.log();
      console.log(fmt.info("Also available via pip:"));
      console.log(`  pip install --upgrade rafter-cli`);
    });

  return cmd;
}
